core/modules/util/threadSafe.h

```diff
@@ -73,21 +73,23 @@ class Flag {
 
     /** Sets flag value to 'val' and returns the old value of flag.
      */
-    virtual T set(T val) {
+    virtual T exchange(T val) {
```

Comment:
If I could use atomic, I would use atomic. In at least one instance I cannot. Using the syntax of atomic where applicable is clear and concise.

The class already exists (written before atomic was available).

Recursive mutex design is up for debate. The recursive_mutex could be avoided, but I doubt it is worth the effort.